pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/PartitionUpsertMetadataManager.java

```diff
@@ -60,128 +64,175 @@
+            if (recordInfo._timestamp >= currentRecordLocation.getTimestamp()) {
               validDocIds.add(recordInfo._docId);
-              return new RecordLocation(segmentName, recordInfo._docId, recordInfo._timestamp, validDocIds);
+              return new RecordLocation(immutableSegment, recordInfo._docId, recordInfo._timestamp);
             } else {
               return currentRecordLocation;
             }
           }
+
+          // The current record is in a different segment
+          // Update the record location when getting a newer timestamp, or the timestamp is the same as the current
+          // timestamp, but the segment has a larger sequence number (the segment is newer than the current segment).
+          if (recordInfo._timestamp > currentRecordLocation.getTimestamp() || (
+              recordInfo._timestamp == currentRecordLocation.getTimestamp() && LLCSegmentName
+                  .isLowLevelConsumerSegmentName(segmentName) && LLCSegmentName
+                  .isLowLevelConsumerSegmentName(currentSegmentName)
+                  && LLCSegmentName.getSequenceNumber(segmentName) > LLCSegmentName
+                  .getSequenceNumber(currentSegmentName))) {
+            assert currentSegment.getValidDocIds() != null;
+            currentSegment.getValidDocIds().remove(currentRecordLocation.getDocId());
+            validDocIds.add(recordInfo._docId);
+            return new RecordLocation(immutableSegment, recordInfo._docId, recordInfo._timestamp);
+          } else {
+            return currentRecordLocation;
+          }
         } else {
           // New primary key
           validDocIds.add(recordInfo._docId);
-          return new RecordLocation(segmentName, recordInfo._docId, recordInfo._timestamp, validDocIds);
+          return new RecordLocation(immutableSegment, recordInfo._docId, recordInfo._timestamp);
         }
       });
     }
     // Update metrics
     _serverMetrics.setValueOfPartitionGauge(_tableNameWithType, _partitionId, ServerGauge.UPSERT_PRIMARY_KEYS_COUNT,
         _primaryKeyToRecordLocationMap.size());
-    return validDocIds;
   }
 
   /**
-   * Updates the upsert metadata for a new consumed record in the given consuming segment.
+   * Updates the upsert metadata for a new consumed record in the given consuming segment. Returns the merged record if
+   * partial-upsert is enabled.
    */
-  public void updateRecord(String segmentName, RecordInfo recordInfo, ThreadSafeMutableRoaringBitmap validDocIds) {
+  public GenericRow updateRecord(IndexSegment segment, RecordInfo recordInfo, GenericRow record) {
+    // For partial-upsert, need to ensure all previous records are loaded before inserting new records.
+    if (_partialUpsertHandler != null) {
+      while (!_partialUpsertHandler.isAllSegmentsLoaded()) {
+        LOGGER
+            .info("Sleeping 1 second waiting for all segments loaded for partial-upsert table: {}", _tableNameWithType);
```

Comment:
is it possible that segment seals during this 1s?

Reply:
This is blocking the ingestion, so it won't seal while waiting